Rubberduck.Refactorings/Abstract/BaseRefactoringWithSuspensionBase.cs

```diff
@@ -0,0 +1,55 @@
+﻿using System.Runtime.ExceptionServices;
+using NLog;
+using Rubberduck.Parsing.Rewriter;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Refactorings.Exceptions;
+
+namespace Rubberduck.Refactorings
+{
+    public abstract class BaseRefactoringWithSuspensionBase<TModel> : BaseRefactoringBase<TModel>
+        where TModel : class, IRefactoringModel
+    {
+        private readonly IParseManager _parseManager;
+
+        private readonly Logger _logger;
+
+        protected BaseRefactoringWithSuspensionBase(IParseManager parseManager, IRewritingManager rewritingManager)
+            : base(rewritingManager)
+        {
+            _parseManager = parseManager;
+            _logger = LogManager.GetLogger(GetType().FullName);
+        }
+
+        protected abstract bool RequiresSuspension(TModel model);
+
+        public override void Refactor(TModel model)
+        {
+            if (RequiresSuspension(model))
+            {
+                RefactorWithSuspendedParser(model);
+            }
+            else
+            {
+                base.Refactor(model);
```

Comment:
If suspension is still optional for a refactoring that inherits this class, then why couldn't parser suspension just be an option for the base refactoring class?

Reply:
I principle, one could have a virtual `RequiresSuspension` on the base class that defaults to always to true and have the supsension implementation there. However, what I like better about this implementation is that it clearly separated both the handling of suspension and of rewrite sessions and the refactorings using suspension and those that do not.

Reply:
BTW, suspesion is not meant to be completely optional for refactorings inheriting from this class. It would make little sense to derive from it ans always return `false`. However, it has to be possible to only suspend in some cases to support refactorings like `RenameRefactoring` that really only need suspension in special cases.

Reply:
Scenario: A refactoring 'X' inherits from `BaseRefactoringBase<TModel>` and makes use of `RenameBaseRefactoring : BaseRefactoringWithSuspensionBase<RenameModel>`  to rename a module.

In the above scenario I could not see how the `RequiresSuspension`  flag of the `RenameBaseRefactoring` (which will be `true`) would be evaluated and handled by 'X' .  Does this simply mean that if a refactoring 'X' calls a code-only  `BaseRefactoringWithSuspensionBase<T>` refactoring, that 'X' must also inherit from `BaseRefactoringWithSuspensionBase<T>`?

Reply:
A refactoring action using suspension should never be used inside another refactoring action. That has the danger of a potential double suspension, which is not supported.

Even if you used such a refactoring inside another refactoring, the suspension would be used to process the call to `Refactor` on the inner refacoring, which performs its own rewrite.

Reply:
> A refactoring action using suspension should never be used inside another refactoring action

I may be missing something.  I thought `RenameBaseRefactoring` was the most likely refactoring to used inside another refactoring.

Reply:
Since it requires suspension, that will not work.

Reply:
Ought the suspension be passed around in a similar manner to the `CancellationToken`? Otherwise it would be difficult to implement a refactoring action where suspension is needed in certain cases.